lib/engine/game/g_18_sj.rb

```diff
@@ -0,0 +1,737 @@
+
+      def main_line_lay?(action)
+        return false unless action
+        return false unless @main_line_hexes.include?(action.hex)
+        return true if @fulfilled_main_line_hexes.include?(main_line_lay(action))
+        return false if main_line_fulfilled_by_other?(action)
+
+        main_line_hex?(action.hex) && connects_main_line?(action.hex)
+      end
+
+      def main_line_fulfilled_by_other?(action)
+        return false if @fulfilled_main_line_hexes.include?(main_line_lay(action))
+
+        @fulfilled_main_line_hexes.each do |info|
+          return true if info[:hex_name] == action.hex.name && info[:tile_name] != action.tile.name
+        end
+        false
+      end
+
+      def main_line_hex?(hex)
+        MAIN_LINE_INFO[hex.name]
+      end
+
+      def connects_main_line?(hex)
+        info = MAIN_LINE_INFO[hex.name]
+        return unless info
+
+        orientation = info[:orientation]
+        edge1 = "#{hex.name}_#{orientation[0]}_0"
+        edge2 = "#{hex.name}_#{orientation[1]}_0"
+        edges = hex.tile.paths.flat_map(&:edges).map(&:id)
+        edges.include?(edge1) && edges.include?(edge2)
+      end
+
+      def remove_main_line_bonus(action)
+        return unless main_line_lay?(action)
+
+        lay = main_line_lay(action)
+        return if @fulfilled_main_line_hexes.include?(lay)
+
+        info = MAIN_LINE_INFO[action.hex.name]
+        @fulfilled_main_line_hexes << lay
+        main_line = info[:main_line]
+        @main_line_built[main_line] = (@main_line_built[main_line] << action.hex.name)
+        return if @main_line_built[main_line].size < MAIN_LINE_COUNT[main_line]
+
+        @log << "-- Main line #{MAIN_LINE_DESCRIPTION[main_line]} completed!"
+        @log << 'Removes icons for main line'
+        remove_icons(@main_line_built[main_line], [main_line])
+      end
+
+      def main_line_lay(action)
+        { hex_name: action.hex.name, tile_name: action.tile.name }
+      end
+
+      def current_cert_limit
+        certs_per_player =
+          case @corporations.reject(&:closed?).size
+          when 10
+            { 2 => 39, 3 => 26, 4 => 20, 5 => 16, 6 => 13 }
+          when 9
+            { 2 => 35, 3 => 23, 4 => 18, 5 => 14, 6 => 12 }
+          when 8
+            { 2 => 30, 3 => 20, 4 => 15, 5 => 12, 6 => 10 }
+          when 7
+            { 2 => 26, 3 => 17, 4 => 13, 5 => 11, 6 => 9 }
+          else
+            game_error("No cert limit defined for #{@corporations.length} corporations")
+          end.freeze
+        set_cert_limit = certs_per_player[@players.size]
+        game_error("No cert limit defined for #{@players.size} players") unless set_cert_limit
+        set_cert_limit
+      end
+
+      def nationalize_major(major)
+        @log << "#{major.name} is nationalized"
+
+        remove_reservation(major)
+        transfer_home_token(@sj, major)
+        transfer_non_home_tokens(@sj, major)
+
+        major.trains.each do |train|
+          minor.remove_train(train)
```

Comment:
corporations can no longer remove train, you need to calll game

what is trains.delete here?

Reply:
I was surprised when I saw that code - what ever would minor be? It turns out to be some copy-pasted code, so minor is undefined here. Now it does not crash..... and that is because you can only be nationalized if corporation is train less, so that piece of code will never be called!

So I removed the loop on line 535.